patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch

```diff
@@ -0,0 +1,695 @@
+-  // instead of multithreaded (MTA). Usually this means device_info_winrt is
+-  // being called from the UI thread.
+-  THR((HRESULT)ro_initialize_wrapper_);
+-
+-  THR(AssureDeviceInformation());
+-
+-Cleanup:
+-  return hr;
+-}
++DeviceInfoWinRTInternal::DeviceInfoWinRTInternal() {}
+ 
+ HRESULT DeviceInfoWinRTInternal::GetDeviceInformationCollection(
+     IVectorView<DeviceInformation*>** device_collection) {
+   HRESULT hr;
++  ComPtr<IActivationFactory> activation_factory;
++  ComPtr<IDeviceInformationStatics> device_info_statics;
+   ComPtr<IAsyncOperation<DeviceInformationCollection*>>
+       async_op_device_info_collection;
+ 
++  // Get the object containing the DeviceInformation static methods.
++  THR(GetActivationFactory(
++      HStringReference(
++          RuntimeClass_Windows_Devices_Enumeration_DeviceInformation)
++          .Get(),
++      &activation_factory));
++
++  THR(activation_factory.As<IDeviceInformationStatics>(&device_info_statics));
++
+   // Call FindAllAsync and then start the async operation.
+-  THR(device_info_statics_->FindAllAsyncDeviceClass(
++  THR(device_info_statics->FindAllAsyncDeviceClass(
+       DeviceClass_VideoCapture, &async_op_device_info_collection));
+ 
+   // Block and suspend thread until the async operation finishes or timeouts.
+@@ -166,9 +139,6 @@ HRESULT DeviceInfoWinRTInternal::GetNumberOfDevices(uint32_t* device_count) {
+   HRESULT hr;
+   ComPtr<IVectorView<DeviceInformation*>> device_info_collection;
+ 
+-  // Assures class has been properly initialized.
+-  THR(AssureInitialized());
+-
+   THR(GetDeviceInformationCollection(
+       device_info_collection.ReleaseAndGetAddressOf()));
+   THR(device_info_collection->get_Size(device_count));
+@@ -190,9 +160,6 @@ HRESULT DeviceInfoWinRTInternal::GetDeviceName(
+   ComPtr<IVectorView<DeviceInformation*>> device_info_collection;
+   ComPtr<IDeviceInformation> device_info;
+ 
+-  // Assures class has been properly initialized.
+-  THR(AssureInitialized());
+-
+   // Gets the device information collection synchronously
+   THR(GetDeviceInformationCollection(
+       device_info_collection.ReleaseAndGetAddressOf()));
+@@ -281,41 +248,26 @@ HRESULT DeviceInfoWinRTInternal::CreateCapabilityMap(
+     const wchar_t* device_unique_id,
+     vector<VideoCaptureCapability>* video_capture_capabilities) {
+   HRESULT hr;
+-  ComPtr<IMediaCaptureInitializationSettings> init_settings;
+-  ComPtr<IMediaCaptureInitializationSettings5> init_settings5;
++
+   ComPtr<IMediaCapture> media_capture;
+   ComPtr<IClosable> media_capture_closable;
+   ComPtr<IAsyncAction> async_action;
+   ComPtr<IVideoDeviceController> video_device_controller;
+   ComPtr<IMediaDeviceController> media_device_controller;
+   ComPtr<IVectorView<IMediaEncodingProperties*>> stream_capabilities;
++
++  ComPtr<IAsyncAction> async_action_media_capture;
++  Event event_wait_for_media_capture_async_action;
++  Event event_wait_for_media_capture_async_action_complition;
```

Comment:
This change has been updated with the "Cascade of ifs" instead of "Jump on Fail". Although "Cascade of ifs" drops the THR macro and its goto, it doesn't drop the requirement of declaring variable at the beginning of the scope.
Declaring variable near to its use would require the use of the "Nested ifs" or "Throw on fail" patterns. "Nested ifs" is not adequate for this code base due its long chain of API calls. "Throw on fail" can't be used because we can't have exceptions in this code (Google guideline).

Reply:
Should be fine now - can be improved IMHO, but consider the following as nitpicking and optional.

Technically "cascade of ifs" does not require declaring all variables at the beginning - you can declare them just above the `if(SUCCEEDED)` block where they are used, which isn't perfect but is better IMHO. It can be argued that having symmetrical blocks for initializations and cleanups is less error-prone ofc.

More importantly, this is (luckily) C++ code, and predominantly (or exclusively - haven't checked all the cases) uses `ComPtr`, so we can arguably do better than the MS docs examples (which are C and use raw pointers needing manual cleanup). I'd say this pattern is better:
```cpp
HRESULT hr;

ComPtr<A> a;
hr = DoSomething(&a);
if (FAILED(hr)) {
  return hr;
}

// Here we do not propagate the error code so can be shorter
ComPtr<B> b;
if (FAILED(DoSomethingElse(&b))) {
  return E_FAIL;
}

if (SomeCondition()) {
  // Variables do not need to be declared at root scope.
  ComPtr<C> c;
  hr = DoSomethingElseEntirely(&c);
  if (FAILED(hr)) {
    return hr;
  }
}
```
Seems the most C++-idiomatic way to me. Thoughts?